packages/flutter/lib/src/material/colors.dart

```diff
@@ -88,6 +90,75 @@ class MaterialAccentColor extends ColorSwatch<int> {
   Color get shade700 => this[700];
 }
 
+/// Signature for the function that returns a (potentially different) color
+/// based on a given set of states.
+typedef ColorResolver = Color Function(Set<MaterialState> states);
+
+/// Defines a [Color] that can react to changes in the state of a Material
+/// component, based on a given set of [MaterialState]s.
+///
+/// This is useful for preserving the accessibility of text in different states
+/// of a component. For example, in a [FlatButton] with blue text, the text will
+/// become more difficult to read when the button is hovered, focused, or pressed,
+/// because the contrast ratio between the button and the text will decrease. To
+/// solve this, you can use [MaterialStateColor] to make the text darker when the
+/// [FlatButton] is hovered, focused, or pressed.
+///
+/// The [resolve] member is the callback that will be used to get the color
+/// in a given context.
+///
+/// This should only be used as parameters when they are documented to take
+/// [MaterialStatefulColor], otherwise only the default state will be used.
+///
+/// Example:
+/// ```dart
+/// Color getTextColor(Set<MaterialSet> states) {
+///   final interactiveStates = <MaterialState>{
+///     MaterialState.pressed,
+///     MaterialState.hovered,
+///     MaterialState.focused,
+///   };
+///   if(states.any(interactiveStates.contains) {
+///     return Colors.blue[900];
+///   }
+///   return Colors.blue[600];
+/// }
+///
+/// FlatButton(
+///   ...
+///   textColor: MaterialStateColor(getTextColor),
+/// ),
+/// ```
+class MaterialStateColor extends Color {
```

Comment:
Instead of requiring developers to define MSCs with a callback, make it optional:

```
abstract class MaterialStateColor extends Color {
  const MaterialStateColor();
  factory MaterialStateColor.from(ColorResolver callback) => _MaterialStateColorFrom(callback);
  Color resolve(Set<MaterialState> states);
}
```

This way a (const even) subclass of MaterialStateColor could be defined. Callers that find the callback version convenient would write `MaterialStateColor.from(myResolveCallback)`.

Reply:
Oh good idea, that way, if they wanted to, developers could subclass `MaterialStateColor` and use it in a `const` context.

Only minor changes I made to your suggestion were:
* Have `MaterialStateColor` constructor take a `int defaultValue` parameter, so that it can pass it too the `super` constructor (required to make it a `const` constructor).
* Called the factory method `resolveWith(ColorResolver callback)`.